### PR TITLE
🌱 Add omitempty to optional status fields

### LINKED ISSUE #7096: v1beta1#ClusterStatus: some fields are missing ",omitempty"
Hello, it appears most fields in `ClusterStatus` type have omitempty json tag. https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1beta1#ClusterStatus except for `ControlPlaneReady` and `InfrastructureReady`.

This prevents us from zeroing the `.Status` field on `ClusterObject` (which also has `,omitempty`) and have a clean json result. https://pkg.go.dev/sigs.k8s.io/cluster-api/api/v1beta1#Cluster

Is this intentional?
/kind support
/area api
cc: @reetasingh

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7096

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -381,7 +381,11 @@ This project follows the [Kubernetes API conventions](https://github.com/kuberne
 * If a field is required (for our controllers to work) and has a default value specified via OpenAPI schema, but we don't want to force users to set the field, we have to mark the field as optional. Otherwise, the client-side kubectl OpenAPI schema validation will force the user to set it even though it would be defaulted on the server-side.
 
 Optional fields have the following properties:
-* An optional field MUST be marked with `+optional` and include an `omitempty` JSON tag.
+* An optional field MUST be marked with `+optional`.
+* In most cases, optional fields should also have the `omitempty` struct tag (the omitempty option 
+  specifies that the field should be omitted from the JSON encoding if the field has an empty value).
+  However, if you want to have different logic for an optional field which is not provided vs. provided
+  with empty values, do not use omitempty.
```
Comment (fabriziopandini) [inline CONTRIBUTING.md:388]: It will be great if we can provide an example here...

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

File: CONTRIBUTING.md (hunk 1 above)
Comment (sbueringer) [reply CONTRIBUTING.md]: I dropped this part. I was just about to explain why this makes sense with MachineDeployment replicas. But apparently the combination +optiona, omitempty + pointer works.

File: api/v1beta1/zz_generated.openapi.go [hunk 2]
```diff
@@ -673,15 +673,13 @@ func schema_sigsk8sio_cluster_api_api_v1beta1_ClusterStatus(ref common.Reference
 					"infrastructureReady": {
 						SchemaProps: spec.SchemaProps{
 							Description: "InfrastructureReady is the state of the infrastructure provider.",
```
Comment (sbueringer) [inline api/v1beta1/zz_generated.openapi.go:675]: Well that's fun:

> Default values can be specified on a field using the +default= tag. Primitives will have their values directly assigned while structs will go through the JSON unmarshalling process. Fields that do not have an omitempty json tag will default to the zero value of their corresponding type if no default is assigned.
> https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

I assume that is something that is only relevant for core types. Our generated OpenAPI types are **not** used by Kubernetes and the apiserver doesn't know anything about our Go types.

So the change in this generated file should have zero impact.

**Comment by sbueringer** [comment]
Just re-pushed the same commit to test if the jobs work with the Kubernetes 1.26 images

File: cmd/clusterctl/client/cluster/mover_test.go [hunk 3]
```diff
@@ -701,6 +701,9 @@ func Test_objectMover_backupTargetObject(t *testing.T) {
 				}
 
 				fmt.Printf("Actual file content %v\n", string(fileContents))
+				if string(fileContents) != expectedFileContents {
+					fmt.Printf("test")
+				}
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover_test.go:706]: Left over?

**Review by fabriziopandini — COMMENTED** [review]
Only one nit

File: cmd/clusterctl/client/cluster/mover_test.go (hunk 3 above)
Comment (sbueringer) [reply cmd/clusterctl/client/cluster/mover_test.go]: Thx! yes :/
Fixed

Comment (sbueringer) [reply cmd/clusterctl/client/cluster/mover_test.go]: Used it as a way to be able to add a break point in the failure case to be able to easily diff with the debugger

**Review by killianmuldoon — APPROVED** [review]
/lgtm

File: controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_types.go [hunk 4]
```diff
@@ -176,29 +176,29 @@ type KubeadmControlPlaneStatus struct {
 	// Total number of non-terminated machines targeted by this control plane
 	// that have the desired template spec.
 	// +optional
-	UpdatedReplicas int32 `json:"updatedReplicas"`
+	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
```
Comment (sbueringer) [inline controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_types.go:179]: /hold

We have to take a look if changing this breaks some of our code. E.g the topology controller reads KCP with unstructured and then has logic on the field being there or not

File: api/v1beta1/cluster_types.go [hunk 5]
```diff
@@ -328,11 +328,11 @@ type ClusterStatus struct {
 
 	// InfrastructureReady is the state of the infrastructure provider.
 	// +optional
-	InfrastructureReady bool `json:"infrastructureReady"`
+	InfrastructureReady bool `json:"infrastructureReady,omitempty"`
 
 	// ControlPlaneReady defines if the control plane is ready.
 	// +optional
-	ControlPlaneReady bool `json:"controlPlaneReady"`
+	ControlPlaneReady bool `json:"controlPlaneReady,omitempty"`
```
Comment (vincepri) [inline api/v1beta1/cluster_types.go:335]: Is this actually a wanted change? Seeing `ControlPlaneReady: false` in a kubectl get would probably give more signal to new users that don't know this field exist, or for existing users that are watching for changes and need to understand what's the state of things.

Comment (vincepri) [reply api/v1beta1/cluster_types.go:335]: Same applies for other, similar fields

Comment (sbueringer) [reply api/v1beta1/cluster_types.go:335]: Thx for bringing this up. I'll bring the discussion back to the issue and close this PR for now

**Comment by sbueringer** [comment]
/close

for now given concerns brought up above. I summarized them on the issue: https://github.com/kubernetes-sigs/cluster-api/issues/7096#issuecomment-1383833097

**Pull request closed without merging**